Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp

```diff
@@ -0,0 +1,384 @@
+{
+  bool any_dirty = false;
+  for (const auto& texture_data : m_dynamic_input_textures)
+  {
+    any_dirty |= GenerateTexture(sec, controller_name, texture_data);
+  }
+
+  if (!any_dirty)
+  {
+    return;
+  }
+  if (Core::GetState() == Core::State::Starting)
+  {
+    return;
+  }
+  if (!g_renderer)
+  {
+    return;
+  }
+  g_renderer->ForceReloadTextures();
+}
+
+bool DynamicInputTextureConfiguration::GenerateTexture(
+    const IniFile::Section* sec, const std::string& controller_name,
+    const DynamicInputTextureData& texture_data) const
+{
+  std::string device_name;
+  if (!sec->Get("Device", &device_name))
+  {
+    return false;
+  }
+
+  auto emulated_controls_iter = texture_data.m_emulated_controllers.find(controller_name);
+  if (emulated_controls_iter == texture_data.m_emulated_controllers.end())
+  {
+    return false;
+  }
+
+  bool device_found = true;
+  auto host_devices_iter = texture_data.m_host_devices.find(device_name);
+  if (host_devices_iter == texture_data.m_host_devices.end())
+  {
+    // If we fail to find our exact device,
+    // it's possible the creator doesn't care (single player game)
+    // and has used a wildcard for any device
+    host_devices_iter = texture_data.m_host_devices.find("");
+
+    if (host_devices_iter == texture_data.m_host_devices.end())
+    {
+      device_found = false;
+    }
+  }
+
+  // Load colorkey image copy
+  auto color_key_image = LoadImage(m_base_path + texture_data.m_colorKeyImageName);
+  bool dirty = false;
+
+  for (auto& [emulated_key, color] : emulated_controls_iter->second)
+  {
+    std::string host_key = "";
+    sec->Get(emulated_key, &host_key);
+
+    const auto color_only_mask = MaskColor(*color_key_image, Pixel{color.r, color.g, color.b, 255});
+    const auto rects = FindRectRegionsOfNeighboringPixels(color_only_mask);
+
+    auto apply_fallback = [&]() {
+      const auto fallback_image = LoadImage(m_base_path + texture_data.m_fallbackImageName);
+
+      for (const auto& rect : rects)
+      {
+        CopyImageRegion(*fallback_image, *color_key_image, rect, rect);
+        dirty = true;
+      }
+    };
+
+    if (!device_found)
+    {
+      // If we get here, that means the controller is set to a
+      // device not exposed to the pack
+      // We still apply the fallback, in case the user
+      // switched devices and wants to see the changes
+      apply_fallback();
+      continue;
+    }
+
+    const auto input_image_iter = host_devices_iter->second.find(host_key);
+    if (input_image_iter == host_devices_iter->second.end())
+    {
+      apply_fallback();
+    }
+    else
+    {
+      const auto host_key_image = LoadImage(m_base_path + input_image_iter->second);
+
+      for (const auto& rect : rects)
+      {
+        if (host_key_image->width == rect.GetWidth() && host_key_image->height == rect.GetHeight())
+        {
+          CopyImageRegion(*host_key_image, *color_key_image,
```

Comment:
`Rect{0, 0, rect.GetWidth(), rect.GetHeight()}` and `dirty = true` can be moved out of the if cases.